### PR TITLE
Code Generation of GC Pointers

### PR DESCRIPTION
**Pull request by jacob-hughes**
Conceptually this is quite straightforward, but the codegen pipeline makes this quite difficult. The key thing to understand is that an LLVM pointer's addresspace determines whether it is a GC pointer or not. This is encoded as part of the Pointer's type, with addrspace(1) meaning it's a GC pointer and addrspace(0) (default when no addrspace is given) being a normal pointer. We identify whether a pointer is GCable in Rust by checking if the value it's pointing to implements the `Scan` trait. This requires us to have access to the original Rust Type when codegenning LLVM pointers, so that we know which addresspace to use.

In a lot of cases, rustc's codegen will create an LLVM Value with an LLVM Type from a MIR value and its Rust Type from the TypeContext. Once this is done, it's not possible to reverse an LLVM Type to get back a Rust Type [0]. This becomes a problem because LLVM pointers are generated with a function: `type_ptr_to(..)` which takes an LLVM Type and generates an LLVM PointerType to it. We can't know if a pointer should be GCable with just the LLVM Type of the pointee, because we can't check if it implements the `Scan` trait.

This means that in the callsites where LLVM pointers are generated, we need to extend them to carry the Rust type of the pointee through if they don't already. In cases where this is not difficult, I've done this. However, there are instances of `type_ptr_to()` being called within the generation of intrinsic methods, such as `memcpy()` which simply create an LLVM pointer from an LLVM variable which was generated at some aribtrary point in the past. There is no easy way here to fetch the original Rust type to check if it implements `Scan`.

I've been playing around with casting the `LLVM Type` in these cases to an `LLVMPointerType` when I know this to be true. This way I can call `getAddressSpace()` and make sure new pointers are created which match. I was hoping to have this done as part of this PR, but unfortunately I've not got it working yet and wanted to get you something to look at before I go any further. I'm hoping that you might be able to see something outside the box which I may have missed by burying myself too deep into the details. This PR does have debug info left in, but I'm hoping this helps you understand the thought process as I don't intend for it to be merged.

### CODE REVIEW

**Comment by vext01** [comment]
OK, I think I understand the problem. All call sites from rustc can be adjusted to pass the correct addrspace, but you don't want to have to change every intrinsic to pass down the addrspaces?

So here's some questions then:
 * Where are intrinsics implemented? LLVM or Rust?
 * How many are there?

**Comment by jacob-hughes** [comment]
> but you don't want to have to change every intrinsic to pass down the addrspaces

Although that's annoying, it's a secondary issue - I imagine I will need to change every intrinsic whether I like it or not. The real problem is that in some of them, there's no easy way to find what the original Rust type was, and so I need a more creative way of knowing when to emit an addrspace(1 | 0) pointer in this case.

> How many are there?

The file changes in this PR show the updated callsite for every call to `type_llvm_ptr`. For some, the additional `rty: Option<Ty<'tcx>>` is `None` intentionally (because it can never be a GC pointer). In other cases, it's `None` just to satisfy the type checker, and I need to find a way to get the rust type to pass it in.

**Comment by vext01** [comment]
>Although that's annoying, it's a secondary issue

Is it though? Presumably you could pass the desired address spaces into the intrinsics as arguments?

> How many are there?

I meant, how many intrinsics are there?

**Comment by jacob-hughes** [comment]
> Is it though? Presumably you could pass the desired address spaces into the intrinsics as arguments?

I'm not sure I understand the question, but what I think you're asking is "Can we pass addresspace information for pointers to the specific intrinsic codegen methods which need them?"

The answer to that would be, in some circumstance, yes this can be done with relative ease. The Rust `Ty` for args to those intrinsics is already available. In other cases, it's not so easy. Take `transmute` as an example: this essentially performs an `LLVMCastPointer` on one `LLVMPointer` to one of a different type. At the point that this intrinsic is called, there is no obvious way to get the `Ty` info for either the source or destination type - it works only with LLVM `Type`s and `Value`s.

This is why I'm working on something that casts the LLVM source `Type` as a `PointerType` if we know it has to be a pointer based on it's use in an intrinsic. This way, we can call `getAddressSpace()` on on that type and make sure the same `addrspace` is used when we emit the destination pointer type.

> I meant, how many intrinsics are there? 

I'm not sure on the exact number, but it's not necessarily intrinsics which make this a problem. It's the use of the pointer codegen function, `type_ptr_to`, in any part of the code gen where `Ty` info is not immediately available. E.g this is also a problem within `pointercast()` and `bitcast()`, which are not intrinsics. 

There are approximately 20 callsites to `type_ptr_to` where there's no immediately obvious way to get a `Ty`, which is why I'm looking at inspecting previously generated `PointerType`s (which generated the correct `addrspace` because they could use `Ty`) as an alternative in those cases.

File: src/librustc_codegen_llvm/builder.rs [hunk 1]
```diff
@@ -1418,14 +1418,14 @@ impl Builder<'a, 'll, 'tcx> {
...
-        let stored_ptr_ty = self.cx.type_ptr_to(stored_ty);
+        let stored_ptr_ty = self.cx.type_ptr_to(stored_ty, None);
 
         assert_eq!(self.cx.type_kind(dest_ptr_ty), TypeKind::Pointer);
 
         if dest_ptr_ty == stored_ptr_ty {
             ptr
         } else {
-            debug!("Type mismatch in store. \
+            eprintln!("Type mismatch in store. \
```
Comment (ltratt) [inline src/librustc_codegen_llvm/builder.rs]: Any reason to change this from `debug` to `eprintln`?

File: src/librustc_codegen_llvm/type_.rs [hunk 2]
```diff
@@ -243,10 +243,25 @@ impl BaseTypeMethods<'tcx> for CodegenCx<'ll, 'tcx> {
         }
     }
 
-    fn type_ptr_to(&self, ty: &'ll Type) -> &'ll Type {
+    fn type_ptr_to(&self, ty: &'ll Type, rty: Option<Ty<'tcx>>) -> &'ll Type {
         assert_ne!(self.type_kind(ty), TypeKind::Function,
                    "don't call ptr_to on function types, use ptr_to_llvm_type on FnType instead");
-        ty.ptr_to()
+
```
Comment (ltratt) [inline src/librustc_codegen_llvm/type_.rs]: Extraneous blank line.

File: src/librustc_codegen_llvm/type_.rs [hunk 3]
```diff
@@ -243,10 +243,25 @@ impl BaseTypeMethods<'tcx> for CodegenCx<'ll, 'tcx> {
...
+            Some(t) => {
+                let mut addr = 0;
+                if self.type_is_scan(t) {
+                    addr = 1;
+                }
+                ty.ptr_to(addr)
+            },
+            None => ty.ptr_to(0) // Default to emitting a non-GC'd ptr
+        };
+        // eprintln!("LLVM Ty: {:?}, Rust Ty: {:?}", ty, rty);
```
Comment (ltratt) [inline src/librustc_codegen_llvm/type_.rs]: Debugging code.

File: src/librustc_codegen_llvm/type_of.rs [hunk 4]
```diff
@@ -241,16 +241,16 @@ impl<'tcx> LayoutLlvmExt<'tcx> for TyLayout<'tcx> {
         if let layout::Abi::Scalar(ref scalar) = self.abi {
             // Use a different cache for scalars because pointers to DSTs
             // can be either fat or thin (data pointers of fat pointers).
-            if let Some(&llty) = cx.scalar_lltypes.borrow().get(&self.ty) {
-                return llty;
-            }
+            // if let Some(&llty) = cx.scalar_lltypes.borrow().get(&self.ty) {
```
Comment (ltratt) [inline src/librustc_codegen_llvm/type_of.rs]: Debuging code.

**Comment by vext01** [comment]
@ltratt it's not supposed to be reviewed yet ;)

> This PR does have debug info left in, but I'm hoping this helps you understand the thought process as I don't intend for it to be merged.

File: src/librustc_codegen_llvm/builder.rs (hunk 1 above)
Comment (jacob-hughes) [reply src/librustc_codegen_llvm/builder.rs]: I'll change this when I turn this from a draft PR into a real one I would like merging. But the hacky answer is that I get much faster compile times without `debug-assertions` turned on and I just use `eprintln!` for debugging the things I care about.

Comment (ltratt) [reply src/librustc_codegen_llvm/builder.rs]: Fair enough.

**Comment by ltratt** [comment]
Mea culpa!

The blank lines, though... spotting those brings back the glory days before rustfmt removed this source of joy from my life!

**Comment by vext01** [comment]
> There are approximately 20 callsites to type_ptr_to where there's no immediately obvious way to get a Ty, which is why I'm looking at inspecting previously generated PointerTypes 

OK, I'm still trying to get to grips with this. And sorry for being slow.

What I think I need is two concrete examples:
 * One where it's trivial
 * One problem case.

Can you be really clear about what code is from what (LLVM/Rust/..).

Thanks

**Comment by ltratt** [comment]
A secondary question is whether we can still run *some* (clearly not all!) interesting programs (for a very loose definition of "interesting") even when we can't currently inspect all those call sites? If we can, maybe we can come back to this a bit later?

File: src/librustc_codegen_llvm/builder.rs [hunk 5]
```diff
@@ -935,7 +935,7 @@ impl BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
         if flags.contains(MemFlags::NONTEMPORAL) {
             // HACK(nox): This is inefficient but there is no nontemporal memmove.
             let val = self.load(src, src_align);
-            let ptr = self.pointercast(dst, self.type_ptr_to(self.val_ty(val)));
+            let ptr = self.pointercast(dst, self.type_ptr_to(self.val_ty(val), None));
```
Comment (jacob-hughes) [inline src/librustc_codegen_llvm/builder.rs]: @vext01 Here's an example of where it's not immediately obvious how I pass `Ty` information to the `type_ptr_to` as I only have the LLVM `Type` immediately to hand.

File: src/librustc_codegen_ssa/mir/block.rs [hunk 6]
```diff
@@ -952,9 +952,8 @@ impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         if by_ref && !arg.is_indirect() {
             // Have to load the argument, maybe while casting it.
             if let PassMode::Cast(ty) = arg.mode {
-                let addr = bx.pointercast(llval, bx.type_ptr_to(
-                    bx.cast_backend_type(&ty))
-                );
+                let ptr = bx.type_ptr_to(bx.cast_backend_type(&ty), Some(arg.layout.ty));
```
Comment (jacob-hughes) [inline src/librustc_codegen_ssa/mir/block.rs]: @vext01 Passing a `Ty` here is relatively easy, because it can be accessed as a field from the `arg`.

**Comment by ltratt** [comment]
Is the problem here that we need to pass around extra state? Or that it's truly disappeared in a way that we can't pass around?

**Comment by vext01** [comment]
I'm just investigating that now.

**Comment by vext01** [comment]
OK, so if I understand correctly. On occasion values disappear into LLVM (at which point Rust type info is lost) and then later are retrieved with a Rust function called `val_ty`, which looks like this:

```
pub fn val_ty(v: &'ll Value) -> &'ll Type {                                          
    unsafe {                                                                         
        llvm::LLVMTypeOf(v)                                                          
    }                                                                                
}
```
Since the address space isn't stored alongside the LLVM type, we have no way to recover it.

Am I right Jake?

**Comment by jacob-hughes** [comment]
> Am I right Jake?

Almost! If the LLVM Type is a `PointerType` subclass, then we _can_ recover the address space by casting it to a `PointerType` and calling `getAddressSpace()`. However, this is not possible if we only have the `Type` of the pointee.

**Comment by vext01** [comment]
OK, this is what I think you need to do Jake:

Make a binding stub to `LLVMGetPointerAddressSpace` (https://llvm.org/doxygen/group__LLVMCCoreTypeSequential.html#ga124b162b69b5def41dde2fda3668cbd9).

Then when you get to code that does a cast, ask the pointer for it's address space, and ensure you carry whichever one it is to the new type you are casting to.

Would that work?

**Comment by vext01** [comment]
We raced.

> However, this is not possible if we only have the Type of the pointee.

So in the case of:
```
let ptr = self.pointercast(dst, self.type_ptr_to(self.val_ty(val), None));
```

Is it not a matter of asking `dst` what address space it resides?

**Comment by vext01** [comment]
> Is it not a matter of asking dst what address space it resides?

I think I meant `src`. This is conceptually a `memcpy(dst, src, ...)`, so we want to carry the address space of the `src` pointer to the `dst` pointer.

**Comment by jacob-hughes** [comment]
I think that will work for this case. The problem lies where we have a `Value` (which is the pointee) and we need to construct a pointer to it in LLVM where either: a) no rust `Ty` info exists for that `Value`; or b) There is no existing LLVM `PointerType` which we can use to extract address space information.

**Comment by vext01** [comment]
Can you point me to such a case and I'll see if I can figure something out.

File: src/librustc_codegen_llvm/va_arg.rs [hunk 7]
```diff
@@ -29,7 +29,7 @@ fn emit_direct_ptr_va_arg(
     slot_size: Align,
     allow_higher_align: bool
 ) -> (&'ll Value, Align) {
-    let va_list_ptr_ty = bx.cx().type_ptr_to(bx.cx.type_i8p());
+    let va_list_ptr_ty = bx.cx().type_ptr_to(bx.cx.type_i8p(), None);
```
Comment (jacob-hughes) [inline src/librustc_codegen_llvm/va_arg.rs]: @vext01 Here's an example (Although I've just had the idea of trying to get the `Ty` via the `OperandRef`.

Comment (vext01) [reply src/librustc_codegen_llvm/va_arg.rs]: Looks like the `layout` field will help?

**Comment by jacob-hughes** [comment]
After much blood, sweat and tears I think I have this working. The PR will have changed quite a bit from here. Instead the `type_ptr_to` functions are now parameterised by an `addrspace` instead of a `Ty`. This means that if, in places where a `Ty` is not available, but the `addrspace` can be determined via other means, we can pass it straight through. An example of this is by querying the `addrspace` of other pointers to the same LLVM `Value`. 

The main fix, however, is to do with how we codegen function args in the caller. Imagine the following Rust example:

```
fn main() {
    let s = S {} // Assume S is bound by the `Scan` trait
    s.meth()
}
```

The (heavily elided) LLVM IR for `main()` which Rustc generates will then look something like this:

```
main:
    %s = alloca %S
    call void @meth(%S* %0) ; reads "arg %0 of type %S*"
    ret void
```

There's a problem here. When we call `S.meth()` in Rust, a pointer to the `alloca %s` will _always_ be created in `addrspace(0)`. Assuming that `S` is a Gc root, it must be passed to `@meth` as an `addrspace(1)` pointer so that the GC can know about it. However, with the codegen context that we have
available when emitting the call, we don't have enough information to know which `addrspace` the pointer should reside in.

This can fix this by comparing the caller's argument types with the actual parameter types of the callee (which already happens in `src/librustc_codegen_llvm/builder.rs::check_call`). If the `addrspace`s are mismatched, then an `AddrSpaceCast` instruction can be performed inserted before the function call.

**Comment by jacob-hughes** [comment]
bors try

File: src/librustc_codegen_ssa/traits/type_.rs [hunk 8]
```diff
@@ -80,6 +82,14 @@ pub trait DerivedTypeMethods<'tcx>: BaseTypeMethods<'tcx> + MiscMethods<'tcx> {
         ty.is_scan(self.tcx(), ty::ParamEnv::reveal_all(), DUMMY_SP)
     }
 
+    fn type_addrspace(&self, ty: Ty<'tcx>) -> usize {
+        if self.type_is_scan(ty) {
+            return 1
```
Comment (ltratt) [inline src/librustc_codegen_ssa/traits/type_.rs]: We don't need the `return`s.

**Comment by jacob-hughes** [comment]
I've upgraded this from a draft PR as, assuming it passes your review, it will be ready to merge.

**Comment by ltratt** [comment]
To be clear, I think you're waiting for comments from @vext01?

**Comment by vext01** [comment]
I've reviewed the new commits, and they LGTM.

Looks like this was a fiddly one.

So let's squash?

**Comment by jacob-hughes** [comment]
Squashed. Reworked the commit message to explain the changes a bit better. Let me know if it makes sense.

**Comment by jacob-hughes** [comment]
I'm going to investigate why bors is sulking now as I can't raise the other PR for the runtime buildbot stuff until this is merged.

**Comment by jacob-hughes** [comment]
bors try

**Comment by vext01** [comment]
Do you want to squash that last commit in?

**Comment by jacob-hughes** [comment]
I'm going to leave it. It wasn't related to this PR and it's a good reminder about how we need to be careful with `#[no_core]`.

**Comment by vext01** [comment]
So, are we ready to invoke bors?

**Comment by jacob-hughes** [comment]
Yes please

**Pull request merged**